packages/open-next/src/adapters/http/responseStreaming.ts

```diff
@@ -0,0 +1,181 @@
+import http from 'node:http';
+import type { ResponseStream } from '../types/aws-lambda.js';
+import { debug, error } from '../logger.js';
+import { getString, headerEnd } from './utils.js';
+
+const HEADERS = Symbol();
+const BODY = Symbol();
+
+export class StreamingServerResponse extends http.ServerResponse {
+  [HEADERS]: Record<string, string> = {};
+  [BODY]: any[] = [];
+  responseStream: ResponseStream;
+  fixHeaders: (headers: Record<string, string>) => void;
+  private _wroteHeader = false;
+  private _done = false;
+  private _hasWritten = false;
+  timer: NodeJS.Timer | null = null;
+
+  get headers() {
+    return this[HEADERS];
+  }
+
+  setHeader(name: string, value: string | number | readonly string[]): this {
+    // @ts-ignore
+    this[HEADERS][name.toLowerCase()] = value;
+    return this;
+  }
+
+  writeHead(
+    statusCode: number,
+    statusMessage?: string | undefined,
+    headers?: http.OutgoingHttpHeaders | http.OutgoingHttpHeader[] | undefined
+  ): this;
+  writeHead(
+    statusCode: number,
+    headers?: http.OutgoingHttpHeaders | http.OutgoingHttpHeader[] | undefined
+  ): this;
+  writeHead(
+    statusCode: unknown,
+    statusMessage?: unknown,
+    headers?: unknown
+  ): this {
+    if (!this._wroteHeader) {
+      this.fixHeaders(this[HEADERS]);
+      try {
+        this.responseStream = awslambda.HttpResponseStream.from(
+          this.responseStream,
+          {
+            statusCode: statusCode as number,
+            headers: this[HEADERS],
+          }
+        );
+        if (!this.timer) {
+          this.createListener();
+        }
+
+        this._wroteHeader = true;
+        debug('writeHead', this[HEADERS]);
+      } catch (e) {
+        this.responseStream.end();
+        error(e);
+      }
+    }
+    return this;
+  }
+
+  // WORKAROUND: It seems that lambda streaming require the event loop to be free
```

Comment:
when does this happen? I wasn't able to find any note neither within Lambda code nor documentation. Is this on our end or AWS's?

Reply:
I've looked at this file https://gist.github.com/conico974/13afd708af20711b97df439b910ceb53#file-index-mjs-L871. That's the only explanation i found as to why this happens, but honestly i don't really understand what's going on here. It might be something on the other end of the internal request that lambda is doing

At first i tried writing directly to the stream, but the streaming never started, it was just waiting for the end to send all the data like a regular request.
What i did is almost the same as now:
- You just comment the listener part
- Everytime there is a `this[BODY].push(chunk)` i called `this.responseStream.write(chunk)`
- And on the `end` method you call `this.responseStream.end()`

There is some really funky stuff going on here https://github.com/sst/open-next/pull/175#issuecomment-1674926950